common/src/main/java/com/emc/mongoose/common/concurrent/DaemonBase.java

```diff
@@ -48,8 +52,14 @@ public final boolean isStarted() {
 	public final void shutdown()
 	throws IllegalStateException {
 		if(stateRef.compareAndSet(State.INITIAL, State.SHUTDOWN)) {
+			synchronized(state) {
```

Comment:
One more question, may it be better to write "if(stateRef.compareAndSet(State.INITIAL, State.SHUTDOWN) || stateRef.compareAndSet(State.STARTED, State.SHUTDOWN))" to avoid the duplication?